### PR TITLE
Development

### PR DESCRIPTION
**Pull request by MarvellousUbani**
# Project Enumerable
## Description
- Created specified methods for the Enumerable Module
-  Tested methods against specified test cases on API Dock

### CODE REVIEW

**Review by Ispirett — CHANGES_REQUESTED** [review]
Hey @MarvellousUbani great work so far :+1: :100: 

Your methods need some work they are not working as they should
Please test against the original methods and make sure they are working as the should.

for example `array.all?(4) should = (array.all?(4)` but it does not work because your does not accept parameters

## Git Flow reminder.

Friendly reminder: You should use Git Flow in this project. However, changing it now would take too much time.
Please make sure you use Github/Git Flow in the next project. You will find instructions in the Patwright lesson description.

File: lib/main.rb [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+  def my_all(arg = nil)
+    unless block_given?
+      if arg.nil?
+        falsy = 0
+        my_each do |x|
+          falsy += 1 if x == false
+        end
+        return falsy.zero?
+      end
+    end
```
Comment (maelfosso) [inline lib/main.rb:77]: Please, it's not working. There is a bug
```ruby
irb(main):010:0> [nil, true, 99].all?
=> false
irb(main):011:0> [nil, true, 99].my_all
=> true
```

File: lib/main.rb [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+    if block_given?
+      count = 0
+      my_each do |x|
+        count += 1 if yield(x) == true
+      end
+      return count
+    end
+
+    return size if arg.nil?
+  end
```
Comment (maelfosso) [inline lib/main.rb:125]: Sorry :disappointed: there is a bug
```ruby
irb(main):031:0> ary.count(2)
=> 2
irb(main):032:0> ary.my_count(2)
=> nil
```

File: lib/main.rb [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+  def my_inject
+    final = self[0]
+    i = 0
+    while i < size
+      final = yield(final, self[i])
+      i += 1
+    end
+    final
+  end
+end
```
Comment (maelfosso) [inline lib/main.rb:177]: Sorry :disappointed: there is a bug
```ruby
irb(main):037:0> ary.inject { |sum, n| sum + n }
=> 9
irb(main):038:0> ary.my_inject { |sum, n| sum + n }
=> 10
irb(main):039:0> ary.inject(9) { |sum, n| sum + n }
=> 18
irb(main):040:0> ary.my_inject(9) { |sum, n| sum + n }
Traceback (most recent call last):
        3: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        2: from (irb):40
        1: from lib/main.rb:109:in `my_inject'
ArgumentError (wrong number of arguments (given 1, expected 0))
irb(main):042:0> ary.inject(1) { |product, n| product * n }
=> 16
irb(main):043:0> ary.my_inject(1) { |product, n| product * n }
Traceback (most recent call last):
        3: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        2: from (irb):43
        1: from lib/main.rb:109:in `my_inject'
ArgumentError (wrong number of arguments (given 1, expected 0))
irb(main):044:0> ary.inject(10) { |product, n| product * n }
=> 160
irb(main):045:0> ary.my_inject(10) { |product, n| product * n }
Traceback (most recent call last):
        3: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        2: from (irb):45
        1: from lib/main.rb:109:in `my_inject'
ArgumentError (wrong number of arguments (given 1, expected 0))
```

**Review by maelfosso — CHANGES_REQUESTED** [review]
Hello @MarvellousUbani 

Thank you for the improvement :clap: 
Sorry but there are still bugs to fix. Please read the comment below.

Kindly make the changes and re-submit a new code review request.
Thanks

File: lib/main.rb (hunk 3 above)
Comment (MarvellousUbani) [reply lib/main.rb:177]: Alright thanks. this is resolved

File: lib/main.rb (hunk 2 above)
Comment (MarvellousUbani) [reply lib/main.rb:125]: Alright resolved as well

File: lib/main.rb (hunk 1 above)
Comment (MarvellousUbani) [reply lib/main.rb:77]: Thanks. This is resolved.

File: lib/main.rb [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+    end
+
+    pos = 0
+    my_each do |x|
+      pos += 1 if yield(x) == true
+    end
+    pos == size
+  end
+
+  def my_any(arg = nil)
```
Comment (samgaco) [inline lib/main.rb]: remember to use the correct name of the methods, for example, you are now trying to emulate [any?](https://apidock.com/ruby/Enumerable/any%3F)

File: lib/main.rb [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+      new_arr << x if yield(x) == true
+    end
+    new_arr
+  end
+
+  def my_all(arg = nil)
+    unless block_given?
+      falsy = 0
+      my_each do |x|
+        falsy += 1 if x == false || x.nil?
```
Comment (samgaco) [inline lib/main.rb]: have into consideration this test case too

` array.all?(Integer)`

and compare it with your method

File: lib/main.rb [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+    my_each do |x|
+      return true if yield(x) == true
+    end
+    false
+  end
+
+  def my_none
+    pos = 0
+    my_each do |x|
+      pos += 1 if yield(x) == false
```
Comment (samgaco) [inline lib/main.rb]: what if we try

```
irb(main):051:0> ["hi", "now"].none?(5)
=> true
```

the method should be able to take an argument as well

you can find all the expected behaviors for the enumerable methods [here](https://ruby-doc.org/core-2.5.0/Enumerable.html)

**Review by samgaco — CHANGES_REQUESTED** [review]
### Hi @MarvellousUbani :high_brightness: 

### Four points to cover:
---------------------------

1. For the **my_each_with_index** , **my_each** , **my_select** and **my_map** methods, a enumerator should be returned if no block is given, example:

```
irb(main):003:0> [1,2,3].select
=> #<Enumerator: [1, 2, 3]:select>
```

but if I try with your code:

```
irb(main):019:0> [1,2,3].my_select
LocalJumpError (no block given (yield))
```

2. **my_all** , **my_any** and **my_none** should be written as **my_all?** , **my_any?** and **my_none?**

3. **my_inject** is not working for the following cases:


* When applying a binary operation specified by a block:

```
irb(main):022:0> (5..50).inject(4){|prod,n| prod*n}
=> 5069015533618896340602101361010794807396273594826752000000000000
```

```
irb(main):042:0> (5..50).my_inject(4){|prod,n| prod*n}
NoMethodError (undefined method `[]' for 5..50:Range)
```




* When a symbol is specified to combine each element of the collection (if we add a sum symbol, all the elements should be summed 

```
irb(main):049:0> [1,2].inject(:+)
=> 3
```

```
irb(main):050:0> [1,2].my_inject(:+)
LocalJumpError (no block given (yield))
```


4. Check all the [expected behaviors of the enumerable ](https://ruby-doc.org/core-2.5.0/Enumerable.html)methods, some methods in your code are not accepting anything but blocks or are not working when no blocks are given as an input.

---------------------------------------------------------------------
Have in mind :arrow_down: 
* **[For your next code review requests]**  Add a short summary of changes in Pull Request description
---------------------------------------------------------------------


* ###### STATUS OF THE PROJECT: REQUESTED CHANGES :stop_sign:

File: lib/main.rb [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+      my_each do |x|
+        num += 1 if x.match(arg)
+      end
+      return num.zero?
+    end
+
+    count.size.zero?
+  end
+
+  def my_count(arg = nil)
```
Comment (bolah2009) [inline lib/main.rb:107]: #### REQUIRED

There is a :bug: with `my_count` method. Please, let's fix this.

```ruby
$ irb
irb(main):001:0> require_relative 'lib/main'
=> true
irb(main):002:0> [1,2,3,4,5].count(4)
=> 1
irb(main):003:0> [1,2,3,4,5].my_count(4)
=> nil
irb(main):004:0> [1,2,3,4,5].count{|i| i > 3}
=> 2
irb(main):005:0> [1,2,3,4,5].my_count{|i| i > 3}
=> 5
```

**Review by bolah2009 — CHANGES_REQUESTED** [review]
## Nice job @MarvellousUbani  :smile: 

#### Awesome progress... :muscle: All methods are working just fine with only one exception. :clap: 

Please see comment below, fix it and submit another code review request. 

- Update PR Description
   update-PR-description

#### :recycle:  Make these changes and submit another code review request. :+1: 
#### Feel free to reach me on slack if you having issues implementing these features. :+1: 

Good luck with your next submission. :star2: 

Cheers
[Bola](https://bolabuari.com/)

File: lib/main.rb (hunk 7 above)
Comment (MarvellousUbani) [reply lib/main.rb:107]: Thanks, boss and this has been resolved.

**Comment by MarvellousUbani** [comment]
# Project Enumerable

## Description
- Created specified methods for the Enumerable Module
- Tested methods against specified test cases on API Dock

File: lib/main.rb [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+    elsif arg.class == Class
+      count = my_select { |x| x.is_a?(arg) }
+    elsif arg.class == Regexp
+      my_each do |x|
+        return true if x.match(arg)
+      end
+    else
+      count = my_select { |x| x == arg }
+    end
+    p count
```
Comment (bolah2009) [inline lib/main.rb]: Let's remove `p` from this method.

**Review by bolah2009 — APPROVED** [review]
## Awesome :clap: 

### It's a :+1: from me.

I would strongly recommend you remove `p` from the method in the comment below before merging

#### :heavy_check_mark:  This project is approved.

**Comment by bolah2009** [comment]
> # Project Enumerable
> ## Description
> * Created specified methods for the Enumerable Module
> * Tested methods against specified test cases on API Dock

### This should be in the PR description above. Kindly edit this (:point_down: ) and put it there. 

marvellouspr

**Pull request merged**